packages/push/src/evaluation/cases.rs

```diff
@@ -0,0 +1,99 @@
+#[derive(Debug)]
+pub struct Case<Input, Output = Input> {
+    pub input: Input,
+    pub output: Output,
+}
+
+#[derive(Debug)]
+pub struct Cases<Input, Output = Input> {
+    cases: Vec<Case<Input, Output>>,
+}
+
+impl<Input, Output> Cases<Input, Output> {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self { cases: Vec::new() }
+    }
+
+    pub fn from_inputs(
```

Comment:
Suggestion: Do we also want a `inputs.with_target(|i| -> Output) -> Cases<Input,Output>` method here?

Reply:
Maybe? Would we impl that on `Vec<Input>`?

Reply:
on T: IntoIterator<Item = Input> I'd say

Reply:
Right – my idea was way too specific.

It does sound like a potentially useful thing to add. I'd thought we might get to it in today's stream, but that didn't happen. It's easy enough that I might just do it now, though.

Reply:
Actually, I got stuck here (so much for "It's easy enough" 😜). I tried:

```rust
impl<T, Input> T
where
    T: IntoIterator<Item = Input>,
{
    pub fn with_target<Output>(
        &self,
        target_function: impl Fn(&Input) -> Output,
    ) -> Cases<Input, Output> {
        Cases::from_inputs(self.into_iter(), target_function)
    }
}
```

but it's not happy because "no nominal type found for inherent implementation". Basically, it doesn't like this super general `T` that it doesn't know very much about.

The "suggested" fix is to create a new wrapper type (maybe `Inputs`?) that implements `IntoIterator`, and then add `with_target` to that new type. That seems like a fair bit of work for a function I'm not sure we'll ever use/need.

Thoughts?

Reply:
you just need a trait, you can't impl a generic directly

```rs
trait WithTarget<Input> {
    fn with_target<Output>(self,  target_fn: impl Fn(&Input) -> Output) -> Cases<Input,Output>;
}

impl<T, Input> WithTarget<Input> for T
where
    T: IntoIterator<Item = Input>,
{
    pub fn with_target<Output>(
        self,
        target_function: impl Fn(&Input) -> Output,
    ) -> Cases<Input, Output> {
        Cases::from_inputs(self.into_iter(), target_function)
    }
}
```